versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/ReferenceLogicNonTxImpl.java

```diff
@@ -0,0 +1,619 @@
+
+    if (!reference.pointer().equals(expectedPointer)) {
+      // A previous deleteReference failed, act as if the first one succeeded, therefore this
+      // one must throw a ReferenceNotFoundException instead of a ReferenceConditionFailedException
+      if (!actAsAlreadyDeleted) {
+        Supplier<StoreIndex<CommitOp>> indexSupplier = createRefsIndexSupplier();
+        reference = maybeRecover(name, reference, indexSupplier);
+        throw new RefConditionFailedException(reference);
+      }
+    }
+
+    if (!actAsAlreadyDeleted) {
+      persist.markReferenceAsDeleted(reference);
+    }
+
+    commitDeleteReference(reference);
+
+    persist.purgeReference(reference);
+
+    if (actAsAlreadyDeleted) {
+      // A previous deleteReference failed, act as if the first one succeeded, therefore this
+      // one would have not found the reference.
+      throw new RefNotFoundException(reference(name, expectedPointer, false));
+    }
+  }
+
+  static final class CommitReferenceResult {
+    final Reference reference;
+    final Kind kind;
+
+    CommitReferenceResult(Reference reference, Kind kind) {
+      this.reference = reference;
+      this.kind = kind;
+    }
+
+    enum Kind {
+      ADDED_TO_INDEX,
+      REF_ROW_EXISTS,
+      REF_ROW_MISSING
+    }
+  }
+
+  @VisibleForTesting // needed to simulate recovery scenarios
+  CommitReferenceResult commitCreateReference(String name, ObjId pointer)
+      throws RetryTimeoutException {
+    try {
+      return commitRetry(
+          persist,
+          (p, retryState) -> {
+            Reference refRefs = p.findReference(REF_REFS.name());
+            long created = p.config().currentTimeMicros();
+            RefObj ref = ref(pointer, created);
+            ObjId refObjId = requireNonNull(ref.id());
+            try {
+              p.storeObj(ref);
+            } catch (ObjTooLargeException e) {
+              throw new RuntimeException(e);
+            }
+
+            StoreKey k = key(name);
+
+            CreateCommit c =
+                newCommitBuilder()
+                    .message("Create reference " + name + " pointing to " + pointer)
+                    .parentCommitId(refRefs.pointer())
+                    .addAdds(commitAdd(k, 0, refObjId, null, null))
+                    .headers(EMPTY_COMMIT_HEADERS)
+                    .commitType(CommitType.INTERNAL)
+                    .build();
+
+            commitReferenceChange(p, refRefs, c);
+
+            return new CommitReferenceResult(reference(name, pointer, false), ADDED_TO_INDEX);
+          });
+    } catch (CommitConflictException e) {
+      checkState(e.conflicts().size() == 1, "Unexpected amount of conflicts %s", e.conflicts());
+
+      CommitConflict conflict = e.conflicts().get(0);
+      checkState(conflict.conflictType() == KEY_EXISTS, "Unexpected conflict type %s", conflict);
+
+      Supplier<StoreIndex<CommitOp>> indexSupplier = createRefsIndexSupplier();
```

Comment:
Should this be using the `REF_REFS` state from the last try (line 381) as opposed to "current" HEAD?

Reply:
Why? It handles the race when another party created the same reference - and if that's the case, it just returns the created one.

Reply:
This piece of code was refactored... I do not recall its original form :)